WordPress/gutenberg
# Modules API: Refactor, tests, and final dependencies array structure

## What?



This is a general refactor of the code, including more test coverage and a few improvements on the API.

Starting from this PR, the dependencies array has the following structure:

> An array of module identifiers of the dependencies of this module. The dependencies can be strings or arrays. If they are arrays, they need an `id` key with the module identifier, and can contain a `type` key with either `static` or `dynamic`. By default, dependencies that don't contain a type are considered static.

Examples:

```js
// foo.js
import x from "static-dependency";
import y from "another-static-dependency";
import z from "yet-another-static-dependency";

const d = await import("dynamic-dependency");
```

```php
gutenberg_register_module( 'foo', '/foo.js', array(
  'static-dependency',
  array(
    'id' => 'another-static-dependency'
  ),
  array(
    'id'   => 'yet-another-static-dependency',
    'type' => 'static'
  ),
  array(
    'id'   => 'dynamic-dependency',
    'type' => 'dynamic'
  ),
) );
```

This opens the door for future additions, like for example dependency versions to support import map scopes:

```php
gutenberg_register_module( 'lib', '/lib.v1.js', array(), '1.2.3' );
gutenberg_register_module( 'lib', '/lib.v2.js', array(), '2.3.4' );

gutenberg_register_module( 'foo', '/foo.js', array( 'lib' ) );
gutenberg_register_module( 'bar', '/bar.js', array( array( 'id' => 'lib', 'version' => '^1.0.0' ) ) );
```

```html
<script type="importmap">
  {
    "imports": {
      "lib": "/lib.v2.js"
    },
    "scopes": {
      "/bar.js": {
        "lib": "/lib.v1.js"
      }
    }
  }
</script>
```

## Why?



Apart from the general improvements here in Gutenberg, I want to open a PR on WordPress Core as soon as this is merged, so more folks can test it out and give feedback.

## How?



I reviewed all the code, comments and tests.

## Testing Instructions






Either add some interactive blocks to your page and check that the modules, preloads and import maps are added correctly to the page, or use the `gutenberg_register_module` and `gutenberg_enqueue_module` functions in your own code.
